contributors/devel/sig-architecture/feature-gates.md

```diff
@@ -1,111 +1,353 @@
+usually manifests as something like:
+
+```
+if disabled(gate) && !newFieldInUse(oldObj) {
+    obj.NewField = nil
+}
+```
+
+#### Validation: fields with no default value
+
+For optional fields without a default value, API validation should *not* check
+the gate.  Instead, the usual pattern for an optional field applies: if the
+field has a value, the value must be validated. This ensures that once an
+object is validated and accepted with a feature gate enabled, subsequent
+changes to the gate will not cause the saved object to fail validation.
+
+#### Validation: fields with a default value
+
+Some fields are optional in the API, but have a default value, which means they
+are effectively required fields as far as the implementation is concerned. Once
+the feature reaches GA and the feature gate is removed, there should never be a
+case where such a field does not have a value.
+
+Validation for these fields usually looks something like:
+
+```
+if obj.NewField == nil {
+    allErrs = append(allErrs, field.Required(...))
+} else {
+    if !newFieldValid(obj.NewField, fldPath.Child("newField")) {
+        allErrs = append(allErrs, field.Invalid(...))
+    }
+}
+```
+
+While a feature gate is active, the validation must also consider the gate's
+state, the API operation in question, and the previous state of the object (in
+case of UPDATE). It would be wrong to assert that the field is required when
+the feature gate is disabled, but also wrong not to require it when the feature
+gate is enabled.
+
+The API's registry code (`pkg/registry/...`) must check the gate before
+validation and pass a flag into the validation logic (usually as a field in an
+"options" struct) to indicate whether the validation code should allow the new
+field or not.
+
+```
+if enabled(gate) || newFieldInUse(oldObj) {
+    options.EnableNewField = true
+}
+ValidateThisObject(obj, oldObj, options)
+```
+
+The validation code then looks something like:
+
+```
+if opts.EnableNewField {
+    if obj.NewField == nil {
+        allErrs = append(allErrs, field.Required(...))
+    } else {
+        if !newFieldValid(obj.NewField, fldPath.Child("newField")) {
+            allErrs = append(allErrs, field.Invalid(...))
+        }
+    }
+}
+```
+
+#### Implementation logic
+
+The implementation of such a feature will usually want to check the gate, and
+neutralize itself when disabled. In general, when a feature gate is disabled,
+the system should behave as if the feature doesn't exist. This may result in
+the API field carrying a value but not actually triggering any functionality.
+This is preferable to a bug in the implementation causing harm that cannot be
+disabled by the gate.
+
+### Features which change an existing API field
+
+For features which don't add a new field, but do change what's allowed in the
+API (e.g. adding a value to an enum or loosening validation), feature gates
+*must* start in alpha. Unlike new fields, which can be remove entirely, we
```

Comment:
I think it's not required to go through alpha for every case of relaxing validation.

https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/2926-job-mutable-scheduling-directives is one example where we did that in the past and we went directly to beta.
In general, I think it depends on what exactly we're relaxing - if we're adding a new possible value the it has to go through Alpha to allow rollback. But if we're allowing for new transitions for updates, Alpha doesn't seems required.

I suggest that we clarify this.

Reply:
The text says "For features which ... change what's allowed in the API (e.g. adding a value to an enum or loosening validation)".  I'll add some words to cover API changes which don't have a "durable" footprint.

Reply:
pushed